gemfire-core/src/main/java/com/gemstone/gemfire/internal/cache/BucketRegion.java

```diff
@@ -1261,13 +1268,106 @@ public void doUnlockForPrimary() {
+    if (lockGIIForSnapshot) {
+      final LogWriterI18n logger = getCache().getLoggerI18n();
+      if (logger.fineEnabled()) {
+        logger.fine("Taking readonly snapshotGIILock on bucket " + this.getName());
+      }
+      snapshotGIILock.attemptLock(LockMode.READ_ONLY, -1, giiReadLockForSIOwner);
+    }
+  }
+
+  public void releaseSnapshotGIIReadLock() {
+    if (lockGIIForSnapshot) {
+      final LogWriterI18n logger = getCache().getLoggerI18n();
+      if (logger.fineEnabled()) {
+        logger.fine("Releasing readonly snapshotGIILock on bucket " + this.getName());
+      }
+      this.snapshotGIILock.releaseLock(LockMode.READ_ONLY, false, giiReadLockForSIOwner);
+    }
+  }
+
+  private MembershipListener giiListener = null;
+
+  public void takeSnapshotGIIWriteLock(MembershipListener listener) {
+    if (lockGIIForSnapshot) {
+      final LogWriterI18n logger = getCache().getLoggerI18n();
+      if (logger.fineEnabled()) {
+        logger.fine("Taking exclusive snapshotGIILock on bucket " + this.getName());
+      }
+      this.snapshotGIILock.attemptLock(LockMode.EX, -1, giiWriteLockForSIOwner);
+      getBucketAdvisor()
+              .addMembershipListenerAndAdviseGeneric(listener);
+      this.giiListener = listener; // Set the listener only after taking the write lock.
+      if (logger.fineEnabled()) {
+        logger.fine("Succesfully took exclusive lock on bucket " + this.getName());
+      }
+    }
+  }
+
+  public void releaseSnapshotGIIWriteLock() {
+    if (lockGIIForSnapshot) {
+      final LogWriterI18n logger = getCache().getLoggerI18n();
+      if (logger.fineEnabled()) {
+        logger.fine("Releasing exclusive snapshotGIILock on bucket " + this.getName());
+      }
+      if (this.snapshotGIILock.getOwnerId(null) == giiWriteLockForSIOwner) {
+        getBucketAdvisor().removeMembershipListener(giiListener);
+        this.giiListener = null;
+        this.snapshotGIILock.releaseLock(LockMode.EX, false, giiWriteLockForSIOwner);
+      }
+      if (logger.fineEnabled()) {
+        logger.fine("Released exclusive snapshotGIILock on bucket " + this.getName());
+      }
+    }
+  }
+
+  private BucketRegion companionRegion;
+
+  private final Object companionRegionSync = new Object();
+
+  /**
+   * Corresponding BucketRegion from shadow table if its a row buffer &
+   * bucket from row buffer if its a shadow table. This method should always be
+   * called in conjuction with PartitionedRegion.needsBatching
+   *
+   * @return
+   */
+  public BucketRegion companionRegion() {
+    if (companionRegion != null) {
+      return companionRegion;
+    }
+    if (!this.getPartitionedRegion().needsBatching() && !this.getPartitionedRegion().
+        getName().toUpperCase().endsWith(StoreCallbacks.SHADOW_TABLE_SUFFIX)) {
+      return companionRegion; // Return null for regions not associated with column tables
+    }
+    synchronized (companionRegionSync) {
+      if (this.getPartitionedRegion().
```

Comment:
Correct. As this fix is a workaround and current SI will work for column tables only we can assume this.